cyten/models/couplings.py

```diff
@@ -0,0 +1,757 @@
+            levels.extend([2 * n + 2, 2 * n + 1])
+            # legs are (before permute) [p0, p0*, p1, p1*, ..., pn, wR, pn*]
+            res = permute_legs(res, domain=['wR'], levels=levels, bend_right=True)
+            W_ = permute_legs(W, codomain=['wL'], bend_right=True)
+            res = tdot(res, W_, 'wR', 'wL')
+        res = squeeze_legs(res, 'wR')
+        codom_labels = [f'p{i}' for i in range(len(self.sites))]
+        dom_labels = [l + '*' for l in codom_labels]
+        # legs are now [p0, p0*, p1, p1*, ...]
+        res = permute_legs(res, codomain=codom_labels, domain=dom_labels,
+                           levels=list(range(2 * len(self.sites))), bend_right=True)
+        return res
+
+    def to_numpy(self) -> np.ndarray:
+        """Convert to a numpy array."""
+        return self.to_tensor().to_numpy()
+
+
+class OnSiteOperator(Coupling):
+    """A (usually hermitian) on-site operator acting on a single :class:`Site`.
+
+    Similar to :class:`Coupling`, but must act on one :class:`Site`.
+
+    TODO examples
+
+    Attributes
+    ----------
+    operator : :class:`SymmetricTensor`
+        Tensor representing the on-site operator with legs ``[p, p*]``, where ``p`` and ``p*`` are
+        the physical space :attr:`sites[0].leg`.
+    sites : list of :class:`Site`
+        Contains the single site that `operator` acts on.
+    factorization : list of :class:`SymmetricTensor`
+        Contains a single tensor corresponding to `operator` with added trivial legs for `wL` and
+        `wR`.
+    name : str, optional
+        A descriptive name that can be used when pretty-printing, to identify the coupling.
+        For example, a Heisenberg coupling is usually initialized with name ``'S.S'``.
+
+    See Also
+    --------
+    :class:`Coupling`
+    """
+
+    def __init__(self, site: DegreeOfFreedom, operator: SymmetricTensor, name: str = None):
+        self.operator = operator
+        W = add_trivial_leg(operator, codomain_pos=0, label='wL')
+        W = add_trivial_leg(W, domain_pos=1, label='wR')
+        # TODO do this for now; discuss how we actually want to do it
+        W.relabel({'p': 'p0', 'p*': 'p0*'})
+        Coupling.__init__(self, sites=[site], factorization=[W], name=name)
+
+    @classmethod
+    def from_tensor(cls, operator: SymmetricTensor, sites: list[DegreeOfFreedom], name: str = None
+                    ) -> OnSiteOperator:
+        assert len(sites) == 1
+        return cls(site=sites[0], operator=operator, name=name)
+
+
+# SPIN COUPLINGS
+
+def spin_spin_coupling(sites: list[SpinDOF], xx: float = None, yy: float = None,
+                       zz: float = None, name: str = 'spin-spin') -> Coupling:
+    """Coupling between two spins.
+
+    Parameters
+    ----------
+    xx, yy, zz : float, optional
+        If given, adds a corresponding term, e.g. ``xx * S_i^x S_j^x`` with the value as prefactor.
+    TODO do we need mixed combinations, like S_i^x S_j^y ?
```

Comment:
maybe give the pre-factors as a 3x3 matrix, maybe with 3-vector? special cases like Heisenberg and Ising should be seperate anyway, so doesnt hurt to be general here?